plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcPageSourceProvider.java

```diff
@@ -0,0 +1,135 @@
+            ConnectorSplit split,
+            ConnectorTableHandle table,
+            List<ColumnHandle> columns,
+            DynamicFilter dynamicFilter)
+    {
+        if (table instanceof JdbcProcedureHandle procedureHandle) {
+            return new RecordPageSource(recordSetProvider.getRecordSet(transaction, session, split, procedureHandle, columns));
+        }
+
+        JdbcTableHandle tableHandle = (JdbcTableHandle) table;
+        Optional<JdbcColumnHandle> mergeRowId = columns.stream()
+                .map(JdbcColumnHandle.class::cast)
+                .filter(column -> column.getColumnName().equalsIgnoreCase(MERGE_ROW_ID))
+                .collect(toOptional());
+        if (mergeRowId.isEmpty()) {
+            return new RecordPageSource(recordSetProvider.getRecordSet(transaction, session, split, tableHandle, columns));
+        }
+
+        List<JdbcColumnHandle> primaryKeys = jdbcClient.getPrimaryKeys(session, tableHandle.getRequiredNamedRelation().getRemoteTableName());
+        List<JdbcColumnHandle> scanColumns = getScanColumns(session, jdbcClient, tableHandle, primaryKeys);
+
+        ImmutableList.Builder<ColumnAdaptation> columnAdaptationsBuilder = ImmutableList.builder();
+        for (ColumnHandle columnHandle : columns) {
+            if (columnHandle.equals(mergeRowId.get())) {
+                columnAdaptationsBuilder.add(buildMergeIdColumnAdaptation(scanColumns, primaryKeys));
+            }
+            else {
+                columnAdaptationsBuilder.add(ColumnAdaptation.sourceColumn(scanColumns.indexOf(columnHandle)));
+            }
+        }
+
+        JdbcTableHandle newTableHandle = new JdbcTableHandle(
+                tableHandle.getRelationHandle(),
+                tableHandle.getConstraint(),
+                tableHandle.getConstraintExpressions(),
+                tableHandle.getSortOrder(),
+                tableHandle.getLimit(),
+                Optional.of(scanColumns),
+                tableHandle.getOtherReferencedTables(),
+                tableHandle.getNextSyntheticColumnId(),
+                tableHandle.getAuthorization(),
+                tableHandle.getUpdateAssignments());
+        return new JdbcPageSource(
+                new RecordPageSource(recordSetProvider.getRecordSet(transaction, session, split, newTableHandle, scanColumns)),
+                columnAdaptationsBuilder.build());
+    }
+
+    private static List<JdbcColumnHandle> getScanColumns(
+            ConnectorSession session,
+            JdbcClient jdbcClient,
+            JdbcTableHandle tableHandle,
+            List<JdbcColumnHandle> primaryKeys)
+    {
+        JdbcTableHandle plainTable = new JdbcTableHandle(
+                tableHandle.getRequiredNamedRelation().getSchemaTableName(),
+                tableHandle.getRequiredNamedRelation().getRemoteTableName(),
+                Optional.empty());
+        List<JdbcColumnHandle> columns = jdbcClient.getColumns(session, plainTable);
```

Comment:
yet another example of using plain table, that could benefit from the refactor in https://github.com/trinodb/trino/pull/23034/commits/fed8ee4ad8556f4ad641a592a2fb55691227d75f#r1799571242